util/util.go

```diff
@@ -685,31 +687,47 @@ func IsKubernetesDefaultToleration(toleration v1.Toleration) bool {
 	return false
 }
 
-func AreIdenticalTolerations(oldTolerations, newTolerations map[string]v1.Toleration) bool {
-	// modified existing tolerations
-	for name := range oldTolerations {
-		if !IsKubernetesDefaultToleration(oldTolerations[name]) && !reflect.DeepEqual(newTolerations[name], oldTolerations[name]) {
-			return false
-		}
+func GetLastAppliedTolerations(obj runtime.Object) (string, error) {
+	objMeta, err := meta.Accessor(obj)
+	if err != nil {
+		return "", fmt.Errorf("cannot get last applied tolerations of invalid object %v: %v", obj, err)
 	}
-	// appended new tolerations
-	for name := range newTolerations {
-		if _, exist := oldTolerations[name]; !exist {
-			return false
-		}
+
+	annos := objMeta.GetAnnotations()
+	if annos == nil {
+		return "", nil
 	}
 
-	return true
+	return annos[LastAppliedTolerationKey], nil
+}
+
+func SetLastAppliedTolerationsAnnotation(lastAppliedTolerations string, obj runtime.Object) error {
+	objMeta, err := meta.Accessor(obj)
+	if err != nil {
+		return fmt.Errorf("cannot set last applied tolerations for invalid object %v: %v", obj, err)
+	}
+	annos := objMeta.GetAnnotations()
+	if annos == nil {
+		annos = map[string]string{}
+	}
+	annos[LastAppliedTolerationKey] = lastAppliedTolerations
+	objMeta.SetAnnotations(annos)
+
+	return nil
 }
 
 func TolerationListToMap(tolerationList []v1.Toleration) map[string]v1.Toleration {
 	res := map[string]v1.Toleration{}
 	for _, t := range tolerationList {
-		res[t.Key] = t
+		res[GetTolerationChecksum(t)] = t
```

Comment:
Add comments to explain why we need to use checksum instead.

Reply:
Awesome catch!!!
Yeah, we need to be careful with v.1.10 upgrade. If we don't we might kill all instance manager pod while the volumes are in the attached state. It will lead to data corruption.

Reply:
Thank you! Fixed

Reply:
with the same t.Key but different operator/effect/value?

Reply:
Sure! Added